src/main/java/frc/robot/subsystems/OuttakeSubsystem.java

```diff
@@ -0,0 +1,164 @@
+// Copyright (c) FIRST and other WPILib contributors.
+// Open Source Software; you can modify and/or share it under the terms of
+// the WPILib BSD license file in the root directory of this project.
+
+package frc.robot.subsystems;
+
+import com.ctre.phoenix.motorcontrol.NeutralMode;
+import com.ctre.phoenix.motorcontrol.TalonFXControlMode;
+import com.ctre.phoenix.motorcontrol.can.TalonFX;
+import edu.wpi.first.math.filter.LinearFilter;
+import edu.wpi.first.wpilibj.Timer;
+import edu.wpi.first.wpilibj.shuffleboard.Shuffleboard;
+import edu.wpi.first.wpilibj.shuffleboard.ShuffleboardTab;
+import edu.wpi.first.wpilibj2.command.SubsystemBase;
+import frc.robot.Constants.Outtake;
+import java.util.Optional;
+
+public class OuttakeSubsystem extends SubsystemBase {
+
+  public static class OuttakeDetails {
+    public final double motorPower;
+    public final Optional<StatorLimit> statorLimit;
+    public final Optional<Double> minTimeSeconds;
+
+    public static class StatorLimit {
+      public final double statorTransitionCurrent;
+
+      public StatorLimit(double statorTransitionCurrent) {
+        this.statorTransitionCurrent = statorTransitionCurrent;
+      }
+    }
+
+    public OuttakeDetails(
+        double motorPower, Optional<StatorLimit> statorLimit, Optional<Double> minTimeSeconds) {
+      this.motorPower = motorPower;
+      this.statorLimit = statorLimit;
+      this.minTimeSeconds = minTimeSeconds;
+    }
+  }
+
+  /** The modes of the drivebase subsystem */
+  public enum Modes {
+    HOLD(Outtake.OuttakeModes.HOLD),
+    INTAKE(Outtake.OuttakeModes.INTAKE),
+    OUTTAKE(Outtake.OuttakeModes.OUTTAKE),
+    OFF(Outtake.OuttakeModes.OFF);
+
+    public final OuttakeDetails outtakeDetails;
+
+    private Modes(OuttakeDetails outtakeDetails) {
+      this.outtakeDetails = outtakeDetails;
+    }
+
+    public boolean modeFinished(
```

Comment:
why is this public? it doesnt make sense as something a command should be allowed to control

Reply:
Might just be my poor programming knowledge, but if the method was private, couldn't the method only be called within the class? Don't we want to expose this method to the advance periodic to see if the mode is finished?

Reply:
This is certainly an edgecase, you don't need to feel bad about not knowing it. Your understanding of private is correct! Making the nested class method private means it can only be accessed within the class... which is your outtakesubsystem.

https://docs.oracle.com/javase/specs/jls/se7/html/jls-6.html#jls-6.6.1:~:text=Otherwise%2C%20if%20the%20member%20or%20constructor%20is%20declared%20private%2C%20then%20access%20is%20permitted%20if%20and%20only%20if%20it%20occurs%20within%20the%20body%20of%20the%20top%20level%20class%20(%C2%A77.6)%20that%20encloses%20the%20declaration%20of%20the%20member%20or%20constructor.

Private is scoping the method to only be callable within the top level class. Nested classes are supposed to be an organizational tool, not an access boundary. If you need to hide implementation details of your nested class from your higher level class, you need to create a new class. For this reason, classes within a top level class share access restrictions with their top level class, and the inverse. Nested class can access all privates, and upper class can do the same--they are essentially one class with some namespacing.

If you ever want to play with access control, try changing the access of a method or member and see what happens! In this case, it works fine, which you can see if you try building after the change.